File: surf-cloud-standalone/src/main/java/dev/slne/surf/cloud/standalone/plugin/StandalonePluginManager.java
package dev.slne.surf.cloud.standalone.plugin;

import dev.slne.surf.cloud.standalone.launcher.Launcher;
import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
import it.unimi.dsi.fastutil.objects.ObjectSet;
import it.unimi.dsi.fastutil.objects.ObjectSets;
import jakarta.annotation.PreDestroy;
import java.nio.file.Path;
import lombok.extern.flogger.Flogger;
import org.springframework.stereotype.Component;

@Component
@Flogger
public class StandalonePluginManager {

  public static final Path PLUGIN_DIRECTORY = Launcher.PLUGIN_DIRECTORY;
  private final ObjectSet<StandalonePlugin> plugins = ObjectSets.synchronize(new ObjectOpenHashSet<>());

  void addPlugin(StandalonePlugin plugin) {
    plugins.add(plugin);
  }

  @PreDestroy
  protected void stopPlugins() {
    for (final StandalonePlugin plugin : plugins) {
      try {
        plugin.stop();
      } catch (Exception e) {
        log.atSevere()
            .withCause(e)
            .log("Failed to stop plugin '%s'", plugin.id());
      }
    }
  }
}

File: surf-cloud-standalone/src/main/kotlin/dev/slne/surf/cloud/standalone/plugin/StandalonePluginManager.kt
package dev.slne.surf.cloud.standalone.plugin

import dev.slne.surf.cloud.api.util.logger
import dev.slne.surf.cloud.api.util.mutableObjectSetOf
import dev.slne.surf.cloud.api.util.synchronize
import java.nio.file.Path


object StandalonePluginManager : Runnable{
    private val log = logger()
    private val plugins = mutableObjectSetOf<StandalonePlugin>().synchronize()

    val PLUGIN_DIRECTORY: Path = Path.of("plugins")

    fun addPlugin(plugin: StandalonePlugin) {
        plugins.add(plugin)
    }

    override fun run() {
        stopPlugins()
    }

    private fun stopPlugins() {
        for (plugin in plugins) {
            try {
                plugin.stop()
            } catch (e: Exception) {
                log.atSevere()
                    .withCause(e)
                    .log("Failed to stop plugin '%s'", plugin.id)
            }
        }
    }
}
